### PR TITLE
Pixels rework

### PR DESCRIPTION
**Pull request by newpavlov**
This PR improves performance by ~15-20%. The main change is in `DecodingImage`, now it is generic over pixel type as was proposed in #23 . Also I've removed elements required for 16 bit images from `ChanceTable`. Either way, support of 16 bit images will require some redesign of the current code, so I don't see any reasons why why should keep it right now. It allows us to drop `fnv` dependency and reduce number of allocations. it also makes `ManiacNode` smaller, now it's size equals to 96 bytes, it's probably possible to make it even smaller, but my quick attempts didn't produce measurable performance improvement.

Also I've added a new `Flif::raw(&self) -> &Box<[u8]>` method in addition to existing `get_raw_pixels`, the latter probably can be marked with `#[deprecated]` attribute. Other than that public API is left intact.

Comparison with `libflif`:
```
test bench_cutout_full_decode ... bench:   6,434,378 ns/iter (+/- 109,096)
test bench_grey_decode        ... bench:   6,076,588 ns/iter (+/- 182,748)
// libflif benchmark
test libflif_cutout_full_decode ... bench:   5,390,452 ns/iter (+/- 151,733)
test libflif_grey_decode        ... bench:   5,060,798 ns/iter (+/- 65,122)
```
In other words with this PR we are ~20% behind in terms of performance.

### CODE REVIEW

File: flif/src/pixels.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+impl Pixel for Rgb {
+    type Channels = RgbChannels;
+    type ChanOrder = [RgbChannels; 3];
+
+    #[inline(always)]
+    fn is_rgba() -> bool { false }
+    #[inline(always)]
+    fn get_value(&self, chan: Self::Channels) -> ColorValue {
```
Comment (dgriffen) [inline flif/src/pixels.rs]: Do you know if the bounds checking gets optimized away here? One advantage of the old `Pixel` type was that there was no bounds checking since index was just a match operation. I would hope the optimizer could see this cant go oob and optimize it though.

File: flif/src/components/transformations/mod.rs [hunk 2]
```diff
@@ -65,7 +65,7 @@ impl ::std::fmt::Display for Transformation {
 }
 
 pub trait Transform: ::std::fmt::Debug + Send + Sync {
-    fn snap(&self, channel: Channel, pixel: &Pixel, value: ColorValue) -> ColorValue {
+    fn snap(&self, channel: Channel, pixel: [i16; 4], value: ColorValue) -> ColorValue {
```
Comment (dgriffen) [inline flif/src/components/transformations/mod.rs]: Could we make these be generic over a `Pixel` instead of an array of `i16`? That would probably make the code a lot cleaner.

If not you should change the `i16` to be `ColorValue`

File: flif/src/pixels.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    fn is_alpha_zero(&self) -> bool;
+
+    /// Return red value if chan is green or blue. Always None for greyscale.
+    fn get_red_pvec(&self, chan: Self::Channels) -> Option<ColorValue>;
+    /// Return green value if chan is blue. Always None for greyscale.
+    fn get_green_pvec(&self, chan: Self::Channels) -> Option<ColorValue>;
+    /// Return alpha value if chan is not alpha. Always None for non-RGBA.
+    fn get_alpha_pvec(&self, chan: Self::Channels) -> Option<ColorValue>;
+
+    fn to_rgba(&self) -> [i16; 4];
```
Comment (dgriffen) [inline flif/src/pixels.rs]: This and the other `i16` in this file should probably be `ColorValue` instead.

File: flif/src/pixels.rs (hunk 1 above)
Comment (newpavlov) [reply flif/src/pixels.rs]: I believe compiler will eliminate bound checks here, as it knows all possible variants. Small godbolt [test](https://godbolt.org/g/KX2cZE) confirms it.

File: flif/src/components/transformations/mod.rs (hunk 2 above)
Comment (newpavlov) [reply flif/src/components/transformations/mod.rs]: I thought about it, but it will require more involved changes to transforms. I'll try to play with it in the following days.

**Comment by newpavlov** [comment]
BTW I can add `libflif` benchmark code if you want it. It can live behind feature gated to avoid linking errors.

**Comment by dgriffen** [comment]
I assume you mean the code that benchmarks this against the official FLIF library? It would be interesting to see but I think that should be managed as a separate PR.

File: flif/src/components/transformations/mod.rs [hunk 4]
```diff
@@ -65,7 +65,7 @@ impl ::std::fmt::Display for Transformation {
 }
 
 pub trait Transform: ::std::fmt::Debug + Send + Sync {
-    fn snap(&self, channel: Channel, pixel: [i16; 4], value: ColorValue) -> ColorValue {
+    fn snap(&self, channel: RgbaChannels, pixel: Rgba, value: ColorValue) -> ColorValue {
```
Comment (dgriffen) [inline flif/src/components/transformations/mod.rs]: This is better than we had before but I still wonder if we can't make this generic over the pixel type. The number of channels is the same for the duration of the decoding so it should be possible to make the `Transform` trait generic. The only snag I forsee right now is that YCoCg only works on 3+ channels but it is not allowed to be used on greyscale images in the first so that should be fixable.
